packages/flutter/lib/src/widgets/animated_cross_fade.dart

```diff
@@ -0,0 +1,207 @@
+/// the two children, and will typically be [Interval] objects. Note that
+/// [firstCurve] is inverted, i.e. it fades out when providing a growing curve
+/// like [Curves.linear]. [sizeCurve] is the curve used to animated between the
+/// size of the fading out child and the size of the fading in child.
+///
+/// In the case where the two children have different sizes, the animation crops
+/// overflowing children during the animation by aligning them to the top.
+/// This means that the bottom will be clipped.
+class AnimatedCrossFade extends StatefulWidget {
+  /// Creates a cross fade animation widget.
+  ///
+  /// The [duration] of the animation is the same for all components (fade in,
+  /// fade out, and size), and you can pass [Interval]s instead of [Curve]s in
+  /// order to have finer control.
+  AnimatedCrossFade({
+    Key key,
+    this.firstChild,
+    this.secondChild,
+    this.firstCurve: Curves.linear,
+    this.secondCurve: Curves.linear,
+    this.sizeCurve: Curves.linear,
+    @required this.crossFadeState,
+    @required this.duration
+  }) : super(key: key) {
+    assert(this.firstCurve != null);
+    assert(this.secondCurve != null);
+    assert(this.sizeCurve != null);
+  }
+
+  /// The first child. It fades in an out according to [crossFadeState].
+  final Widget firstChild;
+
+  /// The second child. It fades in an out according to [crossFadeState].
+  final Widget secondChild;
+
+  /// The state towards which the widget is animating.
+  final CrossFadeState crossFadeState;
+
+  /// The duration of the whole orchestrated animation.
+  final Duration duration;
+
+  /// The fade curve of the first child. This curve is inverted, i.e. the first
+  /// widget is visible when this curve provides the value 0.0, and is hidden
+  /// when the curve provides the value 1.0.
+  final Curve firstCurve;
+
+  /// The fade curve of the second child.
+  final Curve secondCurve;
+
+  /// The curve of the animation between the two children's sizes.
+  final Curve sizeCurve;
+
+  @override
+  _AnimatedCrossFadeState createState() => new _AnimatedCrossFadeState();
+}
+
+class _AnimatedCrossFadeState extends State<AnimatedCrossFade> {
+  _AnimatedCrossFadeState() : super();
+
+  AnimationController _controller;
+  Animation<double> _firstAnimation;
+  Animation<double> _secondAnimation;
+
+  @override
+  void initState() {
+    super.initState();
+    _controller = new AnimationController(duration: config.duration);
+    _firstAnimation = new Tween<double>(
+      begin: 1.0,
+      end: 0.0
+    ).animate(
+      new CurvedAnimation(
+        parent: _controller,
+        curve: config.firstCurve
+      )
+    );
+    _secondAnimation = new CurvedAnimation(
+      parent: _controller,
+      curve: config.secondCurve
+    );
+  }
+
+  @override
+  void dispose() {
+    _controller.dispose();
+    super.dispose();
+  }
+
+  @override
+  void didUpdateConfig(AnimatedCrossFade oldConfig) {
+    super.didUpdateConfig(oldConfig);
+
```

Comment:
There should be a short-circuit return here, when oldConfig and the new one are equal.